torchmultimodal/modules/layers/transformer_decoder.py

```diff
@@ -0,0 +1,199 @@
+        x = x.view(*x_shape)
+        return x
+
+
+class MLPBlock(MLP):
+    """Transformer MLP block."""
+
+    def __init__(self, in_dim, out_dim, hidden_dims):
+        super().__init__(
+            in_dim, out_dim, hidden_dims, 0.0, activation=SiLU, normalization=None
+        )
+
+
+class TransformerLayer(nn.Module):
+    def __init__(
+        self,
+        shape,
+        embedding_dim,
+        num_attention_heads,
+        num_hidden_layers,
+        dropout,
+        attention_dropout,
+    ):
+        super().__init__()
+        self.pre_attention_norm = nn.LayerNorm(embedding_dim)
+        self.post_attention_dropout = nn.Dropout(dropout)
+        self.attention = MultiHeadAttention(
+            shape,
+            embedding_dim,
+            embedding_dim,
+            num_attention_heads,
+            num_hidden_layers,
+            causal=True,
+            attention_module=FullAttention(shape, True, attention_dropout),
+        )
+        self.pre_mlp_norm = nn.LayerNorm(embedding_dim)
+        self.post_mlp_dropout = nn.Dropout(dropout)
+        self.mlp_block = MLPBlock(embedding_dim, embedding_dim, [embedding_dim * 4])
+
+    def forward(self, x, decode_step, decode_idx):
+        h = self.pre_attention_norm(x)
+        if self.training:
+            h = checkpoint(self.attention, h, h, h, decode_step, decode_idx)
+        else:
+            h = self.attention(h, h, h, decode_step, decode_idx)
+        h = self.post_attention_dropout(h)
+        x = x + h
+
+        h = self.pre_mlp_norm(x)
+        if self.training:
+            h = checkpoint(self.mlp_block, h)
+        else:
+            h = self.mlp_block(h)
+        h = self.post_mlp_dropout(h)
+        x = x + h
+
+        return x
+
+
+class TransformerDecoder(nn.Module):
+    def __init__(
+        self,
+        position_embedding_input,
+        position_embedding_output,
+        num_attention_heads,
+        num_hidden_layers,
+        dropout,
+        attention_dropout,
+    ):
+        super().__init__()
+        if (
+            position_embedding_input.embedding_dim
+            != position_embedding_output.embedding_dim
+        ):
+            raise ValueError(
+                f"The position embedding dimension of the input modality {position_embedding_input.embedding_dim}"
+                f"must be equal to that of the output modality {position_embedding_output.embedding_dim}"
+            )
+
+        self.embedding_dim = position_embedding_input.embedding_dim
+        self.right_shift = RightShift(self.embedding_dim)
+        self.position_embedding_input = position_embedding_input
+        self.position_embedding_output = position_embedding_output
+        self.input_seq_len = self.position_embedding_input.seq_len
+        self.output_seq_len = self.position_embedding_output.seq_len
+        total_len = self.input_seq_len + self.output_seq_len
+
+        self.attention_nets = nn.ModuleList(
```

Comment:
```suggestion
        self.transformer_layers = nn.ModuleList(
```

Reply:
also would suggest using `nn.Sequential` instead of `nn.ModuleList` since you will forward pass through these in order anyway. then you can just call `self.transformer_layers(x)` below.